File: R/flt.morph.R
# -------------------------------------------------------------------------
# Morphological filters: mErode, mDilate, mOpen, mClose, mKernel
 
# Copyright (c) 2006 Oleg Sklyar

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License 
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.          

# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  

# See the GNU General Public License for more details.
# GPL license wording: http://www.gnu.org/licenses/gpl.html

# -------------------------------------------------------------------------

mKernel <- function(size = 5, shape = "round") {
    if (size < 3 || (size / 2 == as.integer(size / 2)))
        stop("kernel size must be an odd number >= 3: [3, 5, 7, ...")
    if(switch(shape, round=, square=FALSE, TRUE))
        stop("available shapes 'round' and 'square'")
    res <- matrix(as.integer(FALSE), size, size, byrow = TRUE)
    cx = as.integer(size / 2) + 1
    if (shape == "round") {
        res[cx,] = as.integer(TRUE)
        res[,cx] = as.integer(TRUE)
        for (i in 1:(cx-1))
            for (j in 1:(cx-1))
                if ((cx - i)^2 + (cx - j)^2 <= (cx - 1)^2) {
                    res[i, j] = as.integer(TRUE)
                    res[size - i + 1, j] = as.integer(TRUE)
                    res[i, size - j + 1] = as.integer(TRUE)
                    res[size - i + 1, size - j + 1] = as.integer(TRUE)
                }
        return(res)    
    }

    # otherwise square
    res[] = as.integer(TRUE)
    return(res)
}
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
.mErode <- function(x, iter = 1, kernel = mKernel(5), modify = TRUE) {
    if (!assert(x))
        stop("Wrong class of argument x, Image expected")
    if (x@rgb)
        stop("Only binary images are supported at the moment: {0},{1}")
    if (!is.integer(kernel) || !is.matrix(kernel))
        stop("kernel must be an integer matrix")
    if (!modify) {
        x <- copy(x)
        # 0 - erode, 1 - dilate
        return(.CallEBImage("erodeDilate", x, kernel, as.integer(iter), as.integer(0))) 
    }
    else # original data modified
        invisible(.CallEBImage("erodeDilate", x, kernel, as.integer(iter), as.integer(0))) 
}
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
mErode <- function(x, iter = 1, kernel = mKernel(5)) {
    .mErode(x, kernel, iter, FALSE)
}
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
.mDilate <- function(x, iter = 1, kernel = mKernel(5), modify = TRUE) {
    if (!assert(x))
        stop("Wrong class of argument x, Image expected")
    if (x@rgb)
        stop("Only binary images are supported at the moment: {0},{1}")
    if (!is.integer(kernel) || !is.matrix(kernel))
        stop("kernel must be an integer matrix")
    if (!modify) {
        x <- copy(x)
        # 0 - erode, 1 - dilate
        return(.CallEBImage("erodeDilate", x, kernel, as.integer(iter), as.integer(1))) 
    }
    else # original data modified
        invisible(.CallEBImage("erodeDilate", x, kernel, as.integer(iter), as.integer(1))) 
}
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
mDilate <- function(x, iter = 1, kernel = mKernel(5)) {
    .mDilate(x, kernel, iter, FALSE)
}
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
.mOpen <- function(x, iter = 1, kernel = mKernel(5), modify = TRUE) {
    if (!assert(x))
        stop("Wrong class of argument x, Image expected")
    if (x@rgb)
        stop("Only binary images are supported at the moment: {0},{1}")
    if (!modify)
        x <- copy(x)
    if (modify)
        invisible(.mDilate(.mErode(x, kernel, iter, TRUE), kernel, iter, TRUE))
    else
        return(.mDilate(.mErode(x, kernel, iter, TRUE), kernel, iter, TRUE))
}    
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
mOpen <- function(x, iter = 1, kernel = mKernel(5)) {
    .mOpen(x, kernel, iter, FALSE)
}
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
.mClose <- function(x, iter = 1, kernel = mKernel(5), modify = TRUE) {
    if (!assert(x))
        stop("Wrong class of argument x, Image expected")
    if (x@rgb)
        stop("Only binary images are supported at the moment: {0},{1}")
    if (!modify)
        x <- copy(x)
    if (modify)
        invisible(.mErode(.mDilate(x, kernel, iter, TRUE), kernel, iter, TRUE))
    else
        return(.mErode(.mDilate(x, kernel, iter, TRUE), kernel, iter, TRUE))
}    
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
mClose <- function(x, iter = 1, kernel = mKernel(5)) {
    .mClose(x, kernel, iter, FALSE)
}

File: src/flt_morph.cpp
/* -------------------------------------------------------------------------
Morphological filters: erode, dilate, open and close
Copyright (c) 2005 Oleg Sklyar
See flt_morph.h for license
------------------------------------------------------------------------- */
#include "flt_morph.h"
#include "indexing.h"

#include <R_ext/Error.h>
#include <iostream>

using namespace std;

/* kernel size must be odd number: 3, 5, 7 etc */
inline bool match(int * kernel, Point &ksize, double * data, Point & dsize, Point & at, double & mismatch) {
    int i, j, xx, yy;
    int kcx = ksize.x / 2;
    int kcy = ksize.y / 2;
    for (i = -kcx; i <= kcx; i++)
        for (j = -kcy; j <= kcy; j++) {
            if (kernel[(i + kcx) + (j + kcy) * ksize.x] == 0) continue;
            xx = at.x + i;
            yy = at.y + j;
            if (xx < 0 || yy < 0 || xx >= dsize.x || yy >= dsize.y) continue;
            if (data[xx + yy * dsize.x] == mismatch) return false;
        }
    return true;
}

/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
// function modifies its argument !!!  
SEXP erodeDilate(SEXP x, SEXP kernel, SEXP iters, SEXP alg) {
    double resetto = 0; // erode - check foreground
    if (INTEGER(alg)[0] != 0)
        resetto = 1.0;  // dilate - check background
    try {
        unsigned int i;
        if (!assertImage(x))
            error("wrong argument class, Image expected");
        if (LOGICAL(GET_SLOT(x, mkString("rgb")))[0])
            error("input must be a binary image");
        Point size(INTEGER(GET_DIM(x))[0], INTEGER(GET_DIM(x))[1]);
        unsigned int nimages = INTEGER(GET_DIM(x))[2];
        int * kern = INTEGER(kernel);
        Point ksize(INTEGER(GET_DIM(kernel))[0], INTEGER(GET_DIM(kernel))[1]);
        int nrepeats = INTEGER(iters)[0];
        Point pt;
        for (i = 0; i < nimages; i++) {
            double * data = &(REAL(x)[i * size.x * size.y]);
            for (int it = 0; it < nrepeats; it++) {
                for (int j = 0; j < size.x * size.y; j++) {
                    if (data[j] == resetto) continue;
                    pt = getpoint(j, size.x);    
                    if (!match(kern, ksize, data, size, pt, resetto)) 
                            data[j] = 0.5;
                }
                for (int j = 0; j < size.x * size.y; j++)
                    if (data[j] == 0.5) data[j] = resetto;
            }
        }        
            
    }
    catch(exception &error_) {
        error(error_.what());
    }
    return x;
}
